### ISSUE TITLE
useSelector rerenders when a tracked observable updates even if the returned value stays the same

### ISSUE THREAD

**Issue by axtgr** [LegendApp/legend-state#196]
Here is a generalized contrived example:

```
useSelector(() => {
  obs$.get()
  return 1
})
```

My expectation after reading [the docs](https://legendapp.com/open-source/state/react-api/#useselector) was for it to never rerender, but it actually renders each time obs$ updates. 

I stumbled upon this when I wanted to render a component only when the length of an observable array has changed, but it rerendered on any update to any element of the observable:

```
useSelector(() => obs$.length)

useMount(() => {
  obs$[0].foo.set(true) // Causes a render
})
```

**Comment by jmeistrich** [comment]
Oh wow, yeah that's very wrong. It seems in the rewrite to use `useSyncExternalStore` I didn't properly test that case of the value not changing. I have a fix that I will test more tomorrow and then release an update.

This is what I've got if you want to try it and see if it looks right and works correctly for you:

```ts
import { computeSelector, isPromise, Selector, trackSelector } from '@legendapp/state';
import React, { useRef } from 'react';
import { useSyncExternalStore } from 'use-sync-external-store/shim';
import { reactGlobals } from './react-globals';
import type { UseSelectorOptions } from './reactInterfaces';

interface SelectorFunctions<T> {
    subscribe: (onStoreChange: () => void) => () => void;
    getSnapshot: () => T;
    setSelector: (newSelector: Selector<T>) => void;
}

function createSelectorFunctions<T>(selector: Selector<T>): SelectorFunctions<T> {
    let current: T;

    let notify: () => void;
    let dispose: (() => void) | undefined;

    const run = () => {
        // Dispose if already listening
        dispose?.();

        // Track the selector
        const { value, dispose: _dispose } = trackSelector(selector, _update);

        dispose = _dispose;

        // Set lastValue and return it
        return (current = value);
    };
    const _update = () => {
        const prev = current;
        // When observable changes run() to make sure we're tracking the latest selector
        // and update the current value
        run();
        if (current !== prev) {
            // Notify React that the value has changed
            notify?.();
        }
    };

    // Set up initial tracking
    run();

    return {
        setSelector: (newSelector: Selector<T>) => {
            // Update the selector with the latest one from React so that it uses the latest
            // version of it when the observable changes
            selector = newSelector;
        },
        subscribe: (onStoreChange: () => void) => {
            // Save notify function to use when observable changes
            notify = onStoreChange;

            return () => {
                // Dispose observable listener on unmount
                dispose?.();
                dispose = undefined;
            };
        },
        // Return the current value
        getSnapshot: () => current,
    };
}

export function useSelector<T>(selector: Selector<T>, options?: UseSelectorOptions): T {
    // Short-circuit to skip creating the hook if the parent component is an observer
    if (reactGlobals.inObserver) {
        return computeSelector(selector);
    }

    const ref = useRef<SelectorFunctions<T>>();
    if (ref.current) {
        ref.current.setSelector(selector);
    } else {
        ref.current = createSelectorFunctions<T>(selector);
    }
    const { subscribe, getSnapshot } = ref.current;

    const value = useSyncExternalStore(subscribe, getSnapshot, getSnapshot);

    // Suspense support
    if (options?.suspense) {
        if (isPromise(value)) {
            if (React.use) {
                React.use(value);
            } else {
                throw value;
            }
        }
    }

    return value;
}
```

**Comment by jmeistrich** [comment]
It turns out it's actually much harder than expected, because useSelector is used to wrap around rendering the Reactive components. So the selector needs to be run while rendering, so it needs to re-render to run it again... I think it might need to special case rendering components to always re-render while regular selectors can have smarter logic. But that might be fragile... So I'll think about it a bit more.

**Comment by jmeistrich** [comment]
This is fixed in 1.11.2. Thank you for the report! That was very important to fix.
